salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+        :param reply_queue_name: optional name of the reply queue
+        :param auto_delete_queue: set the type of queue to "auto_delete" (
+        it will be deleted when the last consumer disappears)
+        :param broadcast: whether to broadcast via fanout exchange
+        :return:
+        """
+
+        publish_queue_name = queue_name if queue_name else self.queue_name
+
+        # See https://www.rabbitmq.com/tutorials/amqp-concepts.html
+        exchange = self._fanout_exchange_name if broadcast else ""
+
+        # For the IPC pattern, set up the "reply-to" header and "correlation_id"
+        properties = pika.BasicProperties()
+        properties.reply_to = reply_queue_name if reply_queue_name else None
+        properties.correlation_id = correlation_id if correlation_id else uuid4().hex
+
+        log.debug("Sending payload to queue [{}] and exchange [{}]: {}. Payload properties: {}"
+                  .format(publish_queue_name, exchange, payload, properties))
+
+        try:
+            # TODO: consider different queue types (durable, quorum, etc.)
+            self._channel.queue_declare(publish_queue_name, auto_delete=auto_delete_queue)
+            if broadcast:
+                log.debug("Binding queue to exchange [{}] and exchange [{}]".format(publish_queue_name, exchange))
+                self._channel.queue_bind(publish_queue_name, exchange, routing_key=publish_queue_name)
+            self._channel.basic_publish(exchange=exchange,
+                                        routing_key=publish_queue_name,
+                                        body=payload,
+                                        properties=properties) # add reply queue to the properties
+            log.debug("Sent payload to queue [{}] and exchange [{}]: {}".format(publish_queue_name, exchange, payload))
+        except:
+            log.exception("Error publishing to queue [{}] and exchange [{}]".format(publish_queue_name, exchange))
+            raise
+
+    def consume(self, queue_name=None, timeout=60):
+        """
+        A non-blocking consume takes the next message off the queue.
+        :param queue_name:
+        :param timeout:
+        :return:
+        """
+        log.debug("Consuming payload on queue [{}]".format(queue_name))
+
+        queue_name = queue_name if queue_name else self.queue_name
+        self._channel.queue_declare(queue_name)
+        (method, properties, body) = next(self._channel.consume(queue=queue_name, inactivity_timeout=timeout))
+        return body
+
+    def publish_reply(self, payload, properties):
+        """
+        Publishes reply ``payload`` to the reply queue. Non-blocking.
+        :param payload: message body
+        :param properties: payload properties/metadata
+        :return:
+        """
+        reply_queue_name = properties.reply_to
+
+        assert reply_queue_name
+
+        log.debug("Sending reply payload on queue [{}]: {}. Payload properties: {}".format(reply_queue_name, payload,
```

Comment:
If this is more than just WIP debugging, it *may* make sense to dump this to trace instead :thinking: Might warrant some further discussion about how noisy things should be. I'd probably check other transport implementations for guidance there.

Reply:
After I posted this review, I changed all log statements in rabbitmq.py to dump as INFO; pika library itself is very noisy and dumps a lot of stuff in DEBUG, to the point that it's hard to make sense of what is happening due to noise. So yes, we need to tweak the log levels before this review is ready, but for now the most useful log level for the POC seems to be INFO.